osu.Server.Queues.ScoreStatisticsProcessor/Processors/UserRankCountProcessor.cs

```diff
@@ -18,76 +19,84 @@ public class UserRankCountProcessor : IProcessor
+                return;
 
-            if (bestScore == null)
-            {
-                updateRankCounts(userStats, score.Rank, revert: false);
+            var bestScore = getBestScore(score, conn, transaction);
+
+            // If there's already another higher score than this one, nothing needs to be done.
+            if (bestScore?.ID != score.ID)
                 return;
-            }
 
-            if (score.TotalScore < bestScore.TotalScore) return;
-            if (score.TotalScore == bestScore.TotalScore && score.ID <= bestScore.ID) return;
+            // If this score is the new best and there's a previous higher score, that score's rank should be removed before we apply the new one.
+            var secondBestScore = getSecondBestScore(score, conn, transaction);
+            if (secondBestScore != null)
+                removeRank(userStats, secondBestScore.Rank);
 
-            updateRankCounts(userStats, bestScore.Rank, revert: true);
-            updateRankCounts(userStats, score.Rank, revert: false);
+            Debug.Assert(bestScore != null);
+            addRank(userStats, bestScore.Rank);
         }
 
         public void ApplyGlobal(SoloScoreInfo score, MySqlConnection conn)
         {
         }
 
-        private static SoloScoreInfo? getBestScore(int userId, int beatmapId, int rulesetId, ulong? excludedScoreId, MySqlConnection conn, MySqlTransaction transaction)
+        private static SoloScoreInfo? getSecondBestScore(SoloScoreInfo score, MySqlConnection conn, MySqlTransaction transaction)
+            => getBestScore(score, conn, transaction, 1);
+
+        private static SoloScoreInfo? getBestScore(SoloScoreInfo score, MySqlConnection conn, MySqlTransaction transaction, int offset = 0)
         {
-            // purpose of join with `process_history` is to only look for already-processed scores
             var rankSource = conn.QueryFirstOrDefault<SoloScore?>(
-                "SELECT * FROM solo_scores `s` "
-                + "JOIN solo_scores_process_history `ph` ON `ph`.`score_id` = `s`.`id` "
-                + "WHERE (@ExcludedScoreId IS NULL OR `s`.`id` != @ExcludedScoreId) "
-                + "AND `s`.`user_id` = @UserId "
-                + "AND `s`.`beatmap_id` = @BeatmapId "
-                + "AND `s`.`ruleset_id` = @RulesetId "
-                + "AND `s`.`preserve` = 1 "
-                + "ORDER BY `s`.`data`->'$.total_score' DESC, `s`.`id` DESC "
-                + "LIMIT 1", new
+                "SELECT * FROM solo_scores WHERE `user_id` = @user_id "
+                + "AND `beatmap_id` = @beatmap_id "
+                + "AND `ruleset_id` = @ruleset_id "
+                // preserve is not flagged on the newly arriving score until it has been completely processed (see logic in `ScoreStatisticsQueueProcessor.cs`)
+                // therefore we need to make an exception here to ensure it's included.
+                + "AND `preserve` = 1 OR `id` = @new_score_id "
```

Comment:
I'm... not actually sure on the correctness of precedence here? The intent seems to be

```suggestion
                + "AND (`preserve` = 1 OR `id` = @new_score_id) "
```

and I'd probably feel much safer with the extra set of parens here even if that's how it ends up being executed...

Reply:
AND comes before OR, so the parenthesis do make a difference.

Reply:
I'm not sure where they went. I had them there when testing the query for correctness.

Reply:
Right, I removed them to test whether they were causing this rider fuckup:

![JetBrains Rider 2023-09-13 at 09 39 52](https://github.com/ppy/osu-queue-score-statistics/assets/191335/f554f343-020b-4d39-8efd-7cc1d0fe2d62)

Just to confirm, all testing was done *with* the parens.